signature/cose/envelope_test.go

```diff
@@ -79,144 +39,177 @@ func TestParseEnvelopeError(t *testing.T) {
+					}
+				})
 
-		t.Run(fmt.Sprintf("with %s scheme when expiry is not present", signingScheme), func(t *testing.T) {
-			signRequest, err := newSignRequest(signingScheme)
-			if err != nil {
-				t.Fatalf("newSignRequest(). Error = %s", err)
-			}
-			signRequest.Expiry = time.Time{}
-			_, err = env.Sign(signRequest)
-			if err != nil {
-				t.Fatalf("sign failed. Error = %s", err)
-			}
-		})
+				t.Run(fmt.Sprintf("with %s scheme, %v keyType, %v keySize when expiry is not present", signingScheme, keyType, size), func(t *testing.T) {
+					signRequest, err := newSignRequest(signingScheme, keyType, size)
+					if err != nil {
+						t.Fatalf("newSignRequest() failed. Error = %s", err)
+					}
+					signRequest.Expiry = time.Time{}
+					_, err = env.Sign(signRequest)
+					if err != nil {
+						t.Fatalf("Sign() failed. Error = %s", err)
+					}
+				})
 
-		t.Run(fmt.Sprintf("with %s scheme when signing agent is not present", signingScheme), func(t *testing.T) {
-			signRequest, err := newSignRequest(signingScheme)
-			if err != nil {
-				t.Fatalf("newSignRequest(). Error = %s", err)
-			}
-			signRequest.SigningAgent = ""
-			_, err = env.Sign(signRequest)
-			if err != nil {
-				t.Fatalf("sign failed. Error = %s", err)
-			}
-		})
+				t.Run(fmt.Sprintf("with %s scheme, %v keyType, %v keySize when signingAgent is not present", signingScheme, keyType, size), func(t *testing.T) {
+					signRequest, err := newSignRequest(signingScheme, keyType, size)
+					if err != nil {
+						t.Fatalf("newSignRequest() failed. Error = %s", err)
+					}
+					signRequest.SigningAgent = ""
+					_, err = env.Sign(signRequest)
+					if err != nil {
+						t.Fatalf("Sign() failed. Error = %s", err)
+					}
+				})
 
-		t.Run(fmt.Sprintf("with %s scheme when extended signed attributes are not present", signingScheme), func(t *testing.T) {
-			signRequest, err := newSignRequest(signingScheme)
-			if err != nil {
-				t.Fatalf("newSignRequest(). Error = %s", err)
-			}
-			signRequest.ExtendedSignedAttributes = nil
-			_, err = env.Sign(signRequest)
-			if err != nil {
-				t.Fatalf("sign failed. Error = %s", err)
+				t.Run(fmt.Sprintf("with %s scheme, %v keyType, %v keySize when extended signed attributes are not present", signingScheme, keyType, size), func(t *testing.T) {
+					signRequest, err := newSignRequest(signingScheme, keyType, size)
+					if err != nil {
+						t.Fatalf("newSignRequest() failed. Error = %s", err)
+					}
+					signRequest.ExtendedSignedAttributes = nil
+					_, err = env.Sign(signRequest)
+					if err != nil {
+						t.Fatalf("Sign() failed. Error = %s", err)
+					}
+				})
 			}
-		})
+		}
 	}
 }
 
-// Tests various error scenarios around generating a COSE envelope
 func TestSignErrors(t *testing.T) {
-	env := NewEnvelope()
-	t.Run("when Payload is absent", func(t *testing.T) {
+	env := envelope{}
+	// Testing getSigner()
+	t.Run("errorLocalSigner: when getSigner has privateKeyError", func(t *testing.T) {
```

Comment:
looks like we have a lot of similar tests, maybe we can try table-driven tests like this: https://yourbasic.org/golang/table-driven-unit-test/

Reply:
Looks like this pattern does not do much help in my case because I need to manipulate different fields of signRequest. Changing signer and SigningScheme do not share the same t.Run(), so it's necessary to create more than one test structs.

Reply:
> looks like we have a lot of similar tests, maybe we can try table-driven tests like this: https://yourbasic.org/golang/table-driven-unit-test/

In my case, I need to manipulate different fields of signRequest. For example, one test changes the signer; another one changes the SigningScheme. To reuse the same t.Run(), it's necessary to include a func field in the test struct, this in some extent decreases the readability of the code (there will be a very long test struct slice with a func embedded in each of the test item).